tokenscan/scanner.go

```diff
@@ -0,0 +1,221 @@
+package tokenscan
+
+import (
+	"context"
+	"math"
+	"math/rand"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/gocql/gocql"
+
+	"github.com/scylladb/gocqlx"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/scylladb/gocqlx/qb"
+)
+
+const (
+	// parallelism = nodesInCluster * coresInNode * 3
+	defaultParallelism = 72
+)
+
+type tokenRange struct {
+	startRange int64
+	endRange   int64
+}
+
+// SessionBuilder creates a gocql.Session
+type SessionBuilder interface {
+	CreateSession() (*gocql.Session, error)
+}
+
+// ApplyFunc is a callback function that is called with each retrieved scan item.
+// The scan is aborted if the ApplyFunc returns false
+type ApplyFunc func(context.Context, map[string]interface{}) bool
+
+// Scanner is a scylla token range scanner
+type Scanner struct {
+	sessionBuilder SessionBuilder
+	tokenRanges    []*tokenRange
+	parallelism    int
+	results        prometheus.Counter
+	errors         prometheus.Counter
+}
+
+// Opt is an option func
+type Opt func(*Scanner)
+
+// WithParallelism configures the scan parallelism
+func WithParallelism(parallelism int) Opt {
+	return func(s *Scanner) {
+		s.parallelism = parallelism
+	}
+}
+
+// New creates a new scylla token range scanner
+func New(sessionBuilder SessionBuilder, opts ...Opt) *Scanner {
+	s := &Scanner{
+		sessionBuilder: sessionBuilder,
+
+		parallelism: defaultParallelism,
+		results: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "scyllascanner_results_total",
+			Help: "Number of items returned by the scylla scanner",
+		}),
+		errors: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "scyllascaller_errors_total",
+			Help: "Number of errors encountered while scanning",
+		}),
+	}
+	for _, o := range opts {
+		o(s)
+	}
+
+	s.tokenRanges = getShuffledTokenRanges(s.parallelism)
+	return s
+}
+
+// Describe implements the prometheus.Collector interface
+func (s *Scanner) Describe(ch chan<- *prometheus.Desc) {
+	s.results.Describe(ch)
+	s.errors.Describe(ch)
+}
+
+// Collect implements the prometheus.Collector interface
+func (s *Scanner) Collect(ch chan<- prometheus.Metric) {
+	s.results.Collect(ch)
+	s.errors.Collect(ch)
+}
+
+// Scan performs a token range scan with the provided select builder
+// The scanned items are passed to the ApplyFunc. If the ApplyFunc returns ErrAbort the scanning process
+// is aborted. All other ApplyFunc errors are just logged
+func (s *Scanner) Scan(
+	ctx context.Context,
+	selectBuilder *qb.SelectBuilder,
+	partitionKeyColumns []string,
```

Comment:
We should pass a struct to bind parameters from, we can pass the additional map to use BindStructMap this way one gets more flexibility on binding.